Update docs for 2.0 (#352)
sunpy/ndcube: docs/coordinates.rst

.. _coordinates::

=============================
Coordinates and ndcube Objects
==============================
In the :ref:`ndcube` section we showed how `~ndcube.NDCube`'s slicing ensures the coordinate transformations remain consistent with the data as it is sliced.  In this section we will discuss the many other ways in which the ndcube classes support the integration of data and its coordinates.

But first let's recreate the data and WCS components of an `~ndcube.NDCube` for use in our demonstration.

.. code-block:: python

  >>> import numpy as np
  >>> import astropy.wcs
  >>> data = np.ones((3, 4, 5))
  >>> wcs_input_dict = {
  ... 'CTYPE1': 'WAVE    ', 'CUNIT1': 'Angstrom', 'CDELT1': 0.2, 'CRPIX1': 0, 'CRVAL1': 10, 'NAXIS1': 5,
  ... 'CTYPE2': 'HPLT-TAN', 'CUNIT2': 'deg', 'CDELT2': 0.5, 'CRPIX2': 2, 'CRVAL2': 0.5, 'NAXIS2': 4,
  ... 'CTYPE3': 'HPLN-TAN', 'CUNIT3': 'deg', 'CDELT3': 0.4, 'CRPIX3': 2, 'CRVAL3': 1, 'NAXIS3': 3}
  >>> input_wcs = astropy.wcs.WCS(wcs_input_dict)

.. _extra_coords::

ExtraCoords
===========

> In the :ref:`ndcube` section we saw that the WCS object stored at `nducbe.NDCube.wcs` contains the primary set of coordinate transformations that describe the data.  However, what if you have alternative or additional coordinates that are not represented by the WCS?  The `ndcube.ExtraCoords` class provides users with a mechanism of attaching such coordinates to their `~ndcube.NDCube` instances.  Let's start by creating a `~astropy.time.Time` object representing the times corresponding to each position along the 1st axis of the data array defined above.

Reviewer: I'd give a list of examples here of what the alternative or additional coordinates are

.. code-block:: python

  >>> from astropy.time import Time, TimeDelta
  >>> base_time = Time('2000-01-01', format='fits', scale='utc')
  >>> timestamps = Time([base_time + TimeDelta(60 * i, format='sec') for i in range(data.shape[0])])

Now let's create an `~ndcube.ExtraCoords` instance and add our time extra coordinate to it.  To do this we need to supply the physical type of the coordinate, the array axis to which is corresponds, and the values of the coordinate.  The number of values should equal the length of the axis.

.. code-block:: python

  >>> from ndcube import ExtraCoords
  >>> my_extra_coords = ExtraCoords()
  >>> my_extra_coords.add_coordinate('time', (2,), timestamps)  # TO DO: Change the mapping to 0 when bug fixed.

An indefinite number of coordinates can be added in this way.  Alternatively, we can generate an `~ndcube.ExtraCoords` object from a WCS.  The names of the coordinates can be accessed via the `~ndcube.ExtraCoords.keys` method.

.. code-block:: python

  >>> my_extra_coords.keys()
  ('time',)

> The values of the coordinates can be accessed via `~ndcube.ExtraCoords.wcs`.  This property generates a WCS object based on the coordinates stored in the `~ndcube.ExtraCoords` object.  It should be noted that because `~ndcube.ExtraCoords` is a stand alone object, its WCS does not necessarily match the data array of an `~ndcube.NDCube`.  To convert the extra coordinates into a WCS directly applicable to an `~ndcube.NDCube`, see the :ref:`combined_wcs` section below.

Reviewer: combined_wcs section not linking correctly in readthedocs

.. code-block:: python

  >>> import astropy.units as u
  >>> my_cube = NDCube(data, input_wcs)
  >>> my_cube.global_coords.add('distance', 'pos.distance', 1 * u.m)

`~ndcube.GlobalCoords` allows multiple coordinates of the same physical type.  Therefore when adding a global coordinate, you must provide a unique coordinate name, its physical time and the coordinate value.  The value of the coordinate can be accessed by indexing the `~ndcube.GlobalCoords` instance with the coordinate name.

.. code-block:: python

  >>> my_cube.global_coords['distance']
  <Quantity 1. m>

The coordinate's physical type can be accessed via the `~ndcube.GlobalCoords.physical_types` `dict` property.

.. code-block:: python

  >>> my_cube.global_coords.physical_types['distance']
  'pos.distance'

> Because `~ndcube.GlobalCoords` inherits from `Mapping`, it contains a number of mixin methods similar to those of `dict`.

Reviewer: `Mapping`?

Author: https://docs.python.org/3.8/library/collections.abc.html#collections.abc.Mapping

        [(5039.92178073, 4.56894119e-02), (5040.03143627, 1.79995614e+03),
         (5040.14110851, 3.59986658e+03), (5040.25079745, 5.39950295e+03)]]>)

If the user wants the world coordinates for all the axes, ``axes`` can be set to ``None``, which
is in fact the default.

.. code-block:: python

  >>> my_cube.axis_world_coords()
  (<SpectralCoord [1.02e-09, 1.04e-09, 1.06e-09, 1.08e-09, 1.10e-09] m>,
   <SkyCoord (Helioprojective: obstime=None, rsun=695700.0 km, observer=None): (Tx, Ty) in arcsec
       [[(2160.07821927, 4.56894119e-02), (2159.96856373, 1.79995614e+03),
         (2159.85889149, 3.59986658e+03), (2159.74920255, 5.39950295e+03)],
        [(3600.        , 4.56905253e-02), (3600.        , 1.80000000e+03),
         (3600.        , 3.59995431e+03), (3600.        , 5.39963453e+03)],
        [(5039.92178073, 4.56894119e-02), (5040.03143627, 1.79995614e+03),
         (5040.14110851, 3.59986658e+03), (5040.25079745, 5.39950295e+03)]]>)

By default `~ndcube.NDCube.axis_world_coords` returns the coordinates at the
center of each pixel. However, the pixel edges can be obtained by setting
the ``edges`` kwarg to ``True``. For example:

.. code-block:: python

>   >>> my_cube.axis_world_coords(edges=True)

Reviewer: nice!

       [[(1440.24341188, -899.79647591), (1440.07895112,  899.95636786),
         (1439.91446531, 2699.84625127), (1439.74995445, 4499.59909505),
         (1439.58541853, 6298.94094507)],
        [(2880.05774973, -899.84032206), (2880.00292413,  900.00022848),
         (2879.94809018, 2699.97783871), (2879.89324788, 4499.81838925),
         (2879.83839723, 6299.24788597)],
        [(4319.94225027, -899.84032206), (4319.99707587,  900.00022848),
         (4320.05190982, 2699.97783871), (4320.10675212, 4499.81838925),
         (4320.16160277, 6299.24788597)],
        [(5759.75658812, -899.79647591), (5759.92104888,  899.95636786),
         (5760.08553469, 2699.84625127), (5760.25004555, 4499.59909505),
         (5760.41458147, 6298.94094507)]]>)

`~ndcube.NDCube.axis_world_coords` also allows the user to pick which WCS object should be used, `ndcube.NDCube.wcs` or `ndcube.NDCube.combined_wcs` by setting the ``wcs=`` keyword.  This means that extra_coords can be retrieved, or not, as the user wishes.

.. code-block:: python

  >>> combined_ccords = my_cube.axis_world_coords(wcs=my_cube.combined_wcs)

Working with Raw Coordinates
............................

> If users would prefer not to deal with high level coordinate objects, they can elect to use `ndcube.NDCube.axis_world_coords_values`.  The API for this method is the same as `~ndcube.NDCube.axis_world_coords`.  The only difference is that `~astropy.units.Quantity` objects are returned, one for each physical type requested.  In the above case this means that there would be seperate `~astropy.units.Quantity` objects for latitude and longitude, but they would both have the same 2-D shape.  The `~astropy.units.Quantity` objects are returned in world order and correspond to the physical types in the `~astropy.wcs.WCS.world_axis_physical_types`.  The `~astropy.units.Quantity` objects do not contain important contextual information, such as reference frame, which is needed to fully interpret the coordinate values.  However for some use cases this level of completeness is not needed.

Reviewer: ```suggestion
If users would prefer not to deal with high level coordinate objects, they can elect to use `ndcube.NDCube.axis_world_coords_values`.  The API for this method is the same as `~ndcube.NDCube.axis_world_coords`.  The only difference is that `~astropy.units.Quantity` objects are returned, one for each physical type requested.  In the above case this means that there would be separate `~astropy.units.Quantity` objects for latitude and longitude, but they would both have the same 2-D shape.  The `~astropy.units.Quantity` objects are returned in world order and correspond to the physical types in the `~astropy.wcs.WCS.world_axis_physical_types`.  The `~astropy.units.Quantity` objects do not contain important contextual information, such as reference frame, which is needed to fully interpret the coordinate values.  However for some use cases this level of completeness is not needed.
```

.. _coordinates:

==========================
Coordinate Transformations
==========================
In this section we will discuss the many other ways in which the ndcube classes support the integration of data and its coordinates.
In doing so we will use the data array and WCS object defined below.

.. code-block:: python

  >>> import astropy.wcs
  >>> import numpy as np

  >>> # Define data array.
  >>> data = np.random.rand(4, 4, 5)

  >>> # Define WCS transformations in an astropy WCS object.
  >>> wcs = astropy.wcs.WCS(naxis=3)
  >>> wcs.wcs.ctype = 'WAVE', 'HPLT-TAN', 'HPLN-TAN'
  >>> wcs.wcs.cunit = 'Angstrom', 'deg', 'deg'
  >>> wcs.wcs.cdelt = 0.2, 0.5, 0.4
  >>> wcs.wcs.crpix = 0, 2, 2
  >>> wcs.wcs.crval = 10, 0.5, 1

.. _extra_coords:

> ExtraCoords

Reviewer 2: I feel like this page should start with a dummies guide to WCS?

.. _coordinates:

==========================
Coordinate Transformations
==========================

Introduction to WCS
===================

To describe the mapping between array elements/pixels and real world coordinates, ndcube heavily leverages the World Coordinate System (WCS) framework, specifically the tools written by AstroPy that implement this framework in Python.
WCS allows a wide variety of projections, rotations and transformations be stored and executed.

> Because it allows coordinates transformations to be stored functionally, rather than in memory-heavy lookup tables, and because it caters for both astronomy-specific coordinate systems (e.g. RA & Dec.) as well as simpler, more common ones (e.g. wavelength), WCS has become the most common coordinate transformation framework in astronomy.

Reviewer: silly question - coordinate transforms being wavelength?

> The foundation of the Astropy WCS implementation is the `~astropy.wcs.WCS` object, which stores the critical information describing the coordinate transformations (e.g. the reference pixel and its corresponding coordinate values, ``crpix`` and ``crval``, and the projection type, ``ctype``).

Author: ```suggestion
The most commonly used WCS implementation in Python is the `~astropy.wcs.WCS` object, which stores the critical information describing the coordinate transformations as described by the FITS data model (e.g. the reference pixel and its corresponding coordinate values, ``CRPIX`` and ``CRVAL``, and the projection type, ``CTYPE`` etc.).
```

It also executes these transformations via methods like `~astropy.wcs.WCS.world_to_pixel` and `~astropy.wcs.WCS.world_to_pixel` which convert between pixel indices and world coordinate values.
However, these methods are independent of the data array and the `~astropy.wcs.WCS` object carries little or no information about the data itself.
That is why the ndcube package is needed.

> Nonetheless, AstroPy's WCS implementation is a crucial pillar of ndcube, as is the more generalized offshoot `gWCS <https://gwcs.readthedocs.io/en/stable/>`_ which provides greater generalization for the loss of some performance.

Reviewer 2: ```suggestion
Nonetheless, Astropy's WCS implementation is a crucial pillar of ndcube, as is the more generalized offshoot `gWCS <https://gwcs.readthedocs.io/en/stable/>`_ which provides greater generalization outside of the FITS data model.
```

> Crucially though for ndcube, both implementations adhere to the `AstroPy WCS API (APE 14) <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_.

Reviewer 2: ```suggestion
Crucially though for ndcube, both implementations adhere to the `Astropy WCS API <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_.
```

> A familiarity with WCS and the AstroPy and gWCS Python implementations will be helpful (although hopefully not essential) in understanding this guide.

Reviewer 2: ```suggestion
A familiarity with WCS and the Astropy and gWCS Python implementations will be helpful (although hopefully not essential) in understanding this guide.
```

Or it may be that the users have an alternative set of coordinates to the primary set at ``.wcs``.
To demonstrate how to use `~ndcube.ExtraCoords`, let's start by creating a `~astropy.time.Time` object representing the time at each location along the first axis of the ``data`` array defined above.

.. code-block:: python

  >>> from astropy.time import Time, TimeDelta
  >>> base_time = Time('2000-01-01', format='fits', scale='utc')
  >>> timestamps = Time([base_time + TimeDelta(60 * i, format='sec') for i in range(data.shape[0])])

Now let's create an `~ndcube.ExtraCoords` instance and add our time extra coordinate to it.
To do this we need to supply the physical type of the coordinate, the array axis to which is corresponds, and the values of the coordinate.
The number of values should equal the length of the axis and the physical type must be a valid `IVOA UCD1+ controlled words <http://www.ivoa.net/documents/REC/UCD/UCDlist-20070402.html>`_ word.
If one does not exist for your coordinate, prepend the type with ``custom:``.

.. code-block:: python

  >>> from ndcube import ExtraCoords
  >>> my_extra_coords = ExtraCoords()

>   >>> my_extra_coords.add_coordinate('time', (2,), timestamps)  # TO DO: Change the mapping to 0 Issue #342 resolved.

Reviewer 2: Also open an issue for this, or add a comment to #342 or we will forget.

```suggestion
  >>> my_extra_coords.add_coordinate('time', (2,), timestamps)  # TODO: Change the mapping to 0 Issue #342 resolved.
```

.. code-block:: python

  >>> my_cube.global_coords.physical_types['distance']
  'pos.distance'

Because `~ndcube.GlobalCoords` inherits from `~collections.abc.Mapping`, it contains a number of mixin methods similar to those of `dict`.

.. code-block:: python

  >>> list(my_cube.global_coords.keys())  # Returns a list of global coordinate names
  ['distance']
  >>> list(my_cube.global_coords.values())  # Returns a list of coordinate values
  [<Quantity 1. m>]
  >>> list(my_cube.global_coords.items())  # Returns a list of (name, value) pairs
  [('distance', <Quantity 1. m>)]

A common use case for `~ndcube.GlobalCoords` is associated with slicing (:ref:`cube_slicing`).
In addition to tracking and updating the `~ndcube.NDCube.wcs` and `~ndcube.NDCube.extra_coords` objects, `~ndcube.NDCube`'s slicing infrastucture also identifies when the array axes to which a coordinate corresponds are dropped.
The values of dropped coordinates at the position where the `~ndcube.NDCube` was sliced are stored in the `astropy.wcs.WCS` instance from where `~ndcube.GlobalCoords` can access and return them.

> .. code-block:: python

Reviewer 2: if only this example actually worked :stuck_out_tongue:

Author: Indeed!

  >>> my_2d_cube = my_cube[:, :, 0]
  >>> my_2d_cube.array_axis_physical_types  # Note the wavelength axis is now gone.
  [('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
   ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon')]

  >>> # The wavelength value at the slicing location is now in the GLobalCoords object.
  >>> list(my_2d_cube.global_coords.keys())  # doctest: +SKIP
  ['distance', 'em.wl']  # doctest: +SKIP
  >>> my_2d_cube.global_coords.physical_types['em.wl']  # doctest: +SKIP
  'em.wl'  # doctest: +SKIP
  >>> my_2d_cube.global_coords['em.wl']  # doctest: +SKIP
  <SpectralCoord 1e-9 m>

.. _cube_coordinates:

> NDCube Coordinates

Reviewer 2: Should this not be the first section in this document? It feels weird going from a WCS description to EC and GC. Would this then EC then GC be a better flow?

Author: Possibly.  Will give it some though before merging.

==================

> WCS objects are a powerful and concise way of storing complex functional coordinate transformations.

Reviewer: is this astropy WCS objects yeah?

For example:

.. code-block:: python

  >>> celestial = my_cube.axis_world_coords('lon')[0]  # Must extract object from returned tuple with [0]
  >>> my_cube.dimensions
  <Quantity [4., 4., 5.] pix>
  >>> celestial.shape
  (4, 4)
  >>> celestial
  <SkyCoord (Helioprojective: obstime=None, rsun=695700.0 km, observer=None): (Tx, Ty) in arcsec
    [[(2160.07821927, 4.56894119e-02), (2159.96856373, 1.79995614e+03),
      (2159.85889149, 3.59986658e+03), (2159.74920255, 5.39950295e+03)],
     [(3600.        , 4.56905253e-02), (3600.        , 1.80000000e+03),
      (3600.        , 3.59995431e+03), (3600.        , 5.39963453e+03)],
     [(5039.92178073, 4.56894119e-02), (5040.03143627, 1.79995614e+03),
      (5040.14110851, 3.59986658e+03), (5040.25079745, 5.39950295e+03)],
     [(6479.70323031, 4.56860725e-02), (6479.92250932, 1.79982456e+03),
      (6480.14182173, 3.59960344e+03), (6480.36116753, 5.39910830e+03)]]>

It is also possible to request more than one axis's world coordinates by setting ``axes`` to an iterable of data axis number and/or axis type strings.

> The coordinate objects are returned in world axis order in accordance with APE 14.

Reviewer 2: ```suggestion
The coordinate objects are returned in world axis order.
```

For example:

.. code-block:: python

  >>> my_cube.axis_world_coords(edges=True)
  (<SpectralCoord [1.01e-09, 1.03e-09, 1.05e-09, 1.07e-09, 1.09e-09, 1.11e-09] m>, <SkyCoord (Helioprojective: obstime=None, rsun=695700.0 km, observer=None): (Tx, Ty) in arcsec
      [[(1440.24341188, -899.79647591), (1440.07895112,  899.95636786),
        (1439.91446531, 2699.84625127), (1439.74995445, 4499.59909505),
        (1439.58541853, 6298.94094507)],
       [(2880.05774973, -899.84032206), (2880.00292413,  900.00022848),
        (2879.94809018, 2699.97783871), (2879.89324788, 4499.81838925),
        (2879.83839723, 6299.24788597)],
       [(4319.94225027, -899.84032206), (4319.99707587,  900.00022848),
        (4320.05190982, 2699.97783871), (4320.10675212, 4499.81838925),
        (4320.16160277, 6299.24788597)],
       [(5759.75658812, -899.79647591), (5759.92104888,  899.95636786),
        (5760.08553469, 2699.84625127), (5760.25004555, 4499.59909505),
        (5760.41458147, 6298.94094507)],
       [(7199.36047891, -899.70880283), (7199.63452676,  899.86866585),
        (7199.90861634, 2699.58313412), (7200.18274766, 4499.1606028 ),
        (7200.45692072, 6298.32719784)]]>)

> `~ndcube.NDCube.axis_world_coords` also allows the user to pick which WCS object should be used, `ndcube.NDCube.wcs` or `ndcube.NDCube.combined_wcs` by setting the ``wcs=`` keyword.

Reviewer 2: you can also just pass `ndcube.extra_coords` to get the EC tables back.

Author: Currently that doesn't work.
